content/what-is-jsx.md

```diff
@@ -0,0 +1,255 @@
+need to be defined. We will go into the specific details about how the complate
+implementation below, but first we want to look at a few more details about the
+JSX implementation.
+
+<!--
+	here I want HTML so that I can define an anchor element to the section and
+	link to it!
+-->
+
+### Interpolation with JavaScript Expressions
+
+Everything which appears in a `{ }` block will be interpreted as a JavaScript
+expression.
+
+```jsx
+<MyComponent title={title}>
+    <p>{description}</p>
+</MyComponent>
+```
+
+will become
+
+```javascript
+createElement(MyComponent, { title: title },
+	createElement("p", null, description));
+```
+
+This means that we can use JavaScript expressions to express conditionals
+
+```jsx
+<MyComponent title={title || "unknown"}>
+    {description && (
+        <p>{description}</p>
+    )}
+</MyComponent>
+```
+
+or for a loop
+
+```jsx
+<List>
+    {items.map(item => (
+        <span>{item}</span>
+    ))}
+</List>
+```
+
+### Special Cases for Boolean Attributes
+
+Attributes without an explicit value in a JSX expression are interpreted as
+having the boolean value `true`. This means that the following two JSX blocks
+are identical:
+
+```jsx
+<button disabled>
+```
+
+```jsx
+<button disabled={true}>
+```
+
+### Fragment provided by JSX Implementation
+
+JSX always expects exactly one root element in an expression, but in some cases
+we want to return several at once. To do this, a JSX implementation will
+provide a `Fragment` component with provides an artificial hierarchy level.
+
+```jsx
+<dl>
+	{items.map(({ term, description }) => (
+		<Fragment>
+			<dt>{term}</dt>
+			<dd>{description}</dd>
+		</Fragment>
+	))}
+</dl>
+```
+
+### Trying it out
+
+You can try out how JSX translation works in the
+[Babel online compiler](https://babeljs.io/repl#?babili=false&browsers=&build=&builtIns=false&spec=false&loose=false&code_lz=DwWQngwg9gtgDlAdgU0QFwARoJZoDbIC8ARABLJ55QYDqUATngCbEB8AUAJDBytX3IYGbHADOAVxjAA9L3Yzw0eElRpWQA&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=module&lineWrap=true&presets=es2015%2Creact%2Cstage-2&prettier=false&targets=&version=7.7.3&externalPlugins=).
+It translates it using the `createElement` function from the React library,
+but we can also tell Babel to use a different `createElement` function from a
+library like complate.
+
+
+The complate JSX implementation
+-------------------------------
+
+Now we want to look at the specifics of the complate JSX implementation.
+
+We saw before that
+
+```jsx
+<MyComponent title="Hello World">
+	<p>lorem ipsum</p>
+<MyComponent>
+```
+
+becomes
+
+```javascript
+createElement(MyComponent, { title: title },
+	createElement("p", null, description));
+```
+
+In order for this to work, the complate `createElement` function needs
+to be available in the JavaScript scope.
+
+`createElement` in complate expects a user defined component (called a macro)
```

Comment:
Should we mention that class components also exist? (And are maybe not supported in complate?)

Reply:
Because complate macros are necessarily stateless (render and forget, if you will), there's never been any need for supporting classes AFAIK.

However, this does make for a good example of how implementations (the "interpretation" step) might differ.

Reply:
The JSX spec doesn't talk about state at all. Indeed, it is an implementation detail of React which one allows state (nowadays, both forms do).

Reply:
Yes, agreed - however, JSX also doesn't care whether the identifier is a function, a class or indeed a string (think `Heading = condition ? "h2" : "h3"; <Heading>`)?

Reply:
Kann man in complate auch Klassen als Makros verwenden? (das habe ich persönlich nie verwendet).

Ich habe aber schon `Heading = condition ? "h2" : "h3"` → `<Heading>` verwendet. 🤔

Reply:
Nope, complate doesn't support classes: Because there's no state or life cycle, that never came up except in relation to other JSX implementations (notably React).